File: 백준 2110번 공유기설치.cpp
#include<iostream>
#include<algorithm>
#include<vector>
using namespace std;
int n,c;
vector<int > v;
int answer;
bool possible(int n1)
{
	int cnt(1);
	int last(v[0]);
	for(auto i: v)
	{
		if(i-last>=n1)
		{
			cnt++;
			last=i;
		}
	}
	if(cnt>=c)
	return true;
	return false;
}
int main()
{
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cin>>n>>c;
	v.resize(n);
	for(int i=0; i<n; i++)
	cin>>v[i];
	auto p=minmax_element(v.begin(),v.end());
	int left(1),right(*p.second-*p.first);
	sort(v.begin(),v.end());
	while(left<=right)
	{
		int mid((left+right)/2);
		if(possible(mid))
		{
			answer=mid;
			left=mid+1;
		}else
			right=mid-1;
	}
	cout<<answer;
}

File: Boj-Kotlin&Java/boj2110/백준 2110번 공유기 설치.kt
package boj2110
lateinit var house: IntArray
var answer=0
fun main() {
    val (n, c) = readLine()!!.split(" ").map { it.toInt() }
    house = IntArray(n)
    for (i in 0 until n) {
        house[i] = readLine()!!.toInt()
    }
    house.sort()

    var left = 1
    var right = house.last()
    while (left <= right) {
        val mid = (left + right) / 2
        if (check(mid,c)) {
            left = mid + 1
        } else {
            right = mid - 1
        }
    }
    println(answer)
}

fun check(mid:Int, c: Int) : Boolean{
    var cnt = 1
    var start = house[0]
    for (i in 1 until house.size) {
        if (house[i] - start >= mid) {
            cnt++
            start = house[i]
        }
    }
    if (cnt >= c) {
        answer = mid
        return true
    }
    return false
}